File: src/app/core/models/showdown.model.ts
import type { Card } from './card.model';
import type { CountingSystem } from './counting-system.model';
import { DEVIATION_INDEX_SYSTEM_ID } from './deviation.model';
import { handTotal, isBlackjack, isBust, isSoftHand } from './hand.model';
import type { RuleSet } from './strategy.model';

// Pure dealer-play and settlement logic for the post-count showdown. The card
// source is injected as a `draw` callback so the same shoe the player just
// counted can deal the hand, while the rules stay trivially unit-testable.

// Fewest cards needed to deal a single-box showdown's opening hand (two player +
// two dealer). The page only offers a showdown, and the component only enables
// "deal another", when at least this many cards remain; hits beyond the opening
// deal are handled gracefully if the shoe runs out.
export const MIN_SHOWDOWN_CARDS = 4;

// How many simultaneous boxes the player may occupy. One dealer plays against
// all of them from the same shoe.
export const MIN_SHOWDOWN_SPOTS = 1;
export const MAX_SHOWDOWN_SPOTS = 3;

// Selectable box counts, for the settings dropdown.
export const SHOWDOWN_SPOT_OPTIONS: readonly number[] = [1, 2, 3];

export function clampSpots(spots: number): number {
  if (!Number.isFinite(spots)) return MIN_SHOWDOWN_SPOTS;
  return Math.min(MAX_SHOWDOWN_SPOTS, Math.max(MIN_SHOWDOWN_SPOTS, Math.round(spots)));
}

// Cards consumed by the opening deal for `spots` boxes: two per box plus the
// dealer's two. Splits and hits draw beyond this and are handled gracefully if
// the shoe runs dry mid-hand.
export function minCardsForSpots(spots: number): number {
  return clampSpots(spots) * 2 + 2;
}

// Whether the count the showdown is carrying can be graded against, and how.
//
// The insurance index is a Hi-Lo true count and the playing indices are the
// same — a level-2 or fractional system reads a different number off the same
// shoe (see `deviationIndexNote`). KO is the one other system the app can grade,
// because its book publishes a running-count schedule of its own. Everything
// else is dealt and settled exactly as before, ungraded, rather than scored
// against numbers that are not its own.
export type CountBasis =
  | { readonly kind: 'true-count'; readonly trueCount: number }
  | { readonly kind: 'running-count'; readonly runningCount: number; readonly insuranceAt: number }
  | { readonly kind: 'ungraded' };

// This system's own true count, or null when it has none to read.
//
// Kept apart from `countBasisFor` because the two questions are different. A
// deviation index is a Hi-Lo number, so only Hi-Lo may be graded against it. A
// bet ramp is the player's own, indexed by whatever true count they are keeping
// — so any balanced system qualifies, exactly as the bet-spread drill allows.
export function trueCountFor(
  system: CountingSystem,
  runningCount: number,
  decksRemaining: number,
): number | null {
  if (!system.balanced || decksRemaining <= 0) return null;
  return Math.trunc(runningCount / decksRemaining);
}

export function countBasisFor(
  system: CountingSystem,
  runningCount: number,
  decksRemaining: number,
): CountBasis {
  if (system.id === DEVIATION_INDEX_SYSTEM_ID) {
    // A shoe dealt to the felt has no decks left to divide by; nothing is
    // dealt from it either, so the value is never actually consumed.
    if (decksRemaining <= 0) return { kind: 'ungraded' };
    return { kind: 'true-count', trueCount: Math.trunc(runningCount / decksRemaining) };
  }
  const insuranceAt = system.keyCounts?.insuranceCount;
  if (insuranceAt !== undefined) {
    return { kind: 'running-count', runningCount, insuranceAt };
  }
  return { kind: 'ungraded' };
}

// Whether the count says to take insurance, or null when this system's count is
// not one the app can grade. The Hi-Lo threshold is not hard-coded here: it is
// read off the deviation chart by the caller, which is where the index lives.
export function insuranceIsCorrect(basis: CountBasis, hiLoThresholdMet: boolean): boolean | null {
  switch (basis.kind) {
    case 'true-count':
      return hiLoThresholdMet;
    case 'running-count':
      return basis.runningCount >= basis.insuranceAt;
    case 'ungraded':
      return null;
  }
}

export type ShowdownOutcome = 'win' | 'lose' | 'push';

export interface Settlement {
  readonly outcome: ShowdownOutcome;
  // The player won (or pushed) with a two-card natural — a real game pays 3:2.
  readonly playerBlackjack: boolean;
  // The dealer held a two-card natural.
  readonly dealerBlackjack: boolean;
}

// Whether the dealer must draw another card. Stands on hard 17 and any total of
// 18+; hits anything 16 and under; hits a soft 17 only under H17.
export function dealerShouldHit(hand: readonly Card[], ruleSet: RuleSet): boolean {
  const total = handTotal(hand);
  if (total < 17) return true;
  if (total > 17) return false;
  return ruleSet === 'H17' && isSoftHand(hand);
}

// Play the dealer's hand to completion from its initial cards, drawing via the
// supplied callback. Stops early if the draw source is exhausted (the caller
// guarantees enough cards for normal play). Returns the final dealer cards.
export function playDealerHand(
  initial: readonly Card[],
  ruleSet: RuleSet,
  draw: () => Card | undefined,
): Card[] {
  const hand = [...initial];
  while (dealerShouldHit(hand, ruleSet)) {
    const card = draw();
    if (!card) break;
    hand.push(card);
  }
  return hand;
}

// Resolve a finished player hand against a finished dealer hand. Order matters:
// naturals settle before bust logic; a player bust loses even if the dealer also
// busts; a dealer bust pays any standing player hand; otherwise the higher total
// wins and equal totals push.
export function settle(
  player: readonly Card[],
  dealer: readonly Card[],
  // Whether the player hand counts as a natural blackjack (3:2). Defaults to a
  // real two-card 21. Split hands pass `false`: a 21 made after splitting is not
  // a natural and pays even money.
  playerNatural: boolean = isBlackjack(player),
): Settlement {
  const playerBlackjack = playerNatural;
  const dealerBlackjack = isBlackjack(dealer);

  if (playerBlackjack || dealerBlackjack) {
    const outcome: ShowdownOutcome =
      playerBlackjack && dealerBlackjack ? 'push' : playerBlackjack ? 'win' : 'lose';
    return { outcome, playerBlackjack, dealerBlackjack };
  }

  if (isBust(player)) return { outcome: 'lose', playerBlackjack, dealerBlackjack };
  if (isBust(dealer)) return { outcome: 'win', playerBlackjack, dealerBlackjack };

  const p = handTotal(player);
  const d = handTotal(dealer);
  const outcome: ShowdownOutcome = p > d ? 'win' : p < d ? 'lose' : 'push';
  return { outcome, playerBlackjack, dealerBlackjack };
}

File: ios/BlackjackTrainer/Views/Screens/ShowdownModel.swift
import Observation
import SwiftUI

/// Post-count showdown: deals a hand from the persistent shoe the player just
/// counted, plays it hit/stand/double/split (re-splits to four hands; split aces
/// take one card), auto-plays the dealer by the active rule set, and settles each
/// hand win/lose/push (3:2 naturals). A box's original two cards may also
/// late-surrender for half the bet when the shared LS rule is enabled (the peek
/// has already settled any dealer natural). With bet sizing on, each round opens
/// on a bet and settles against the persisted bankroll, and a dealer ace offers
/// insurance (half each bet, pays 2:1) before the hole card is checked. Mirrors
/// the web `ShowdownComponent`.
@MainActor
@Observable
final class ShowdownModel {
    enum Phase {
        /// `betting` and `insurance` are only reached when bet sizing is on: the
        /// round waits for a bet before any card is dealt (the point of
        /// practising against the count), and a dealer ace pauses the deal on
        /// the insurance decision before the hole card is checked.
        case betting, insurance, playerTurn, resolved, exhausted
    }

    /// Most a pair can be split to (3 splits → 4 hands), the common casino cap.
    /// The cap is per box: occupying three boxes does not shrink any one box's
    /// splits.
    static let maxHandsPerBox = 4

    var ruleSet: RuleSet
    /// DAS / LS availability from the shared table rules.
    let options: EngineOptions
    /// Boxes to occupy on the opening deal (1–3). One dealer plays against all.
    let spots: Int
    /// Bet sizing: when on, each round opens on a bet and settles against the
    /// persisted bankroll. Off, the showdown is the pure hand tally it has always
    /// been and no chip figure is shown.
    let betting: Bool
    /// The bet each box posts for the coming round.
    private(set) var bet = Bankroll.minBet
    /// Net chips of the round just resolved, for the result line.
    private(set) var roundNet = 0.0
    /// Net chips the round's insurance bet returned, or nil when no insurance
    /// was taken. Settled the moment the hole card is checked.
    private(set) var insuranceNet: Double?
    private(set) var hands: [PlayerHand] = []
    private(set) var activeIndex = 0
    private(set) var dealerCards: [Card] = []
    private(set) var phase: Phase = .playerTurn
    private(set) var remaining = 0

    @ObservationIgnored private let shoe: Shoe
    @ObservationIgnored private let stats: ShowdownStatsStore
    @ObservationIgnored let bankrollStore: BankrollStore
    /// Every card this showdown dealt, in order, handed back on exit so the
    /// counting drill can fold their running-count value into its carried count —
    /// the cards really left the shoe.
    @ObservationIgnored private(set) var dealtCards: [Card] = []

    init(
        shoe: Shoe,
        ruleSet: RuleSet,
        stats: ShowdownStatsStore,
        options: EngineOptions = .default,
        spots: Int = 1,
        betting: Bool = false,
        bankroll: BankrollStore = BankrollStore()
    ) {
        self.shoe = shoe
        self.ruleSet = ruleSet
        self.options = options
        self.stats = stats
        self.spots = Showdown.clampSpots(spots)
        self.betting = betting
        bankrollStore = bankroll
        remaining = shoe.cardsRemaining
        if betting {
            // Size the bet before seeing a card — the count just practised is the
            // only information the decision should rest on.
            bet = clampedBet(Bankroll.minBet)
            phase = .betting
        } else {
            dealHand()
        }
    }

    /// Keep a bet inside both the table minimum and what the bankroll can back
    /// across every occupied box.
    private func clampedBet(_ value: Double) -> Double {
        Bankroll.clampBet(value, bankroll: bankrollStore.bankroll / Double(spots))
    }

    func setBet(_ value: Double) {
        guard phase == .betting else { return }
        bet = clampedBet(value)
    }

    func dealAfterBet() {
        guard phase == .betting else { return }
        dealHand()
    }

    func resetBankroll() {
        bankrollStore.reset()
        bet = clampedBet(Bankroll.minBet)
        phase = .betting
    }

    var showdownStats: ShowdownStats {
        stats.stats
    }

    var winRate: String {
        let current = stats.stats
        guard current.hands > 0 else { return "—" }
        return "\(Int((Double(current.wins) / Double(current.hands) * 100).rounded()))%"
    }

    func onAction(_ action: Action) {
        switch action {
        case .hit: hit()
        case .stand: stand()
        case .double: double()
        case .split: split()
        case .surrender: surrender()
        default: break
        }
    }

    /// Give up the hand for half the bet. It settles as a loss on the spot — the
    /// dealer owes it nothing — so `resolveAll`'s any-live check and the tally
    /// both see a finished hand, and the round moves to the next box.
    private func surrender() {
        guard canSurrender else { return }
        let index = activeIndex
        let bet = hands[index].bet
        hands[index].surrendered = true
        hands[index].done = true
        hands[index].settlement = Settlement(
            outcome: .lose, playerBlackjack: false, dealerBlackjack: false
        )
        stats.record(outcome: .lose, playerBlackjack: false)
        if betting {
            let payout = Bankroll.surrenderForfeit(bet: bet)
            bankrollStore.record(stake: bet, payout: payout)
            roundNet += payout
        }
        activateNextOrResolve()
    }

    /// Between rounds, betting returns to the bet: the count has moved on, so the
    /// spread should be reconsidered rather than silently repeated.
    func dealAnother() {
        if betting {
            guard !bankrollStore.bustedOut else { return }
            // Clear the settled round before the next bet, so nothing on the felt
            // (or in `committed`) belongs to a hand that is already paid.
            hands = []
            dealerCards = []
            bet = clampedBet(bet)
            phase = .betting
            return
        }
        dealHand()
    }

    /// Deal a fresh opening round to every occupied box in casino order: one card
    /// to each box, the dealer's upcard, a second to each box, the dealer's hole
    /// card. Naturals on either side resolve without any player action.
    private func dealHand() {
        guard shoe.cardsRemaining >= Showdown.minCards(forSpots: spots) else {
            phase = .exhausted
            return
        }
        // The guard guarantees the whole opening round, so no draw here is nil.
        var boxes: [[Card]] = Array(repeating: [], count: spots)
        for index in boxes.indices {
            boxes[index].append(contentsOf: [draw()].compactMap(\.self))
        }
        var dealer = [draw()].compactMap(\.self)
        for index in boxes.indices {
            boxes[index].append(contentsOf: [draw()].compactMap(\.self))
        }
        dealer.append(contentsOf: [draw()].compactMap(\.self))

        let posted = betting ? bet : 0
        roundNet = 0
        insuranceNet = nil
        hands = boxes.enumerated().map { PlayerHand(cards: $1, box: $0, bet: posted) }
        dealerCards = dealer
        activeIndex = 0
        phase = .playerTurn

        // A dealer ace pauses on the insurance decision before the peek — but
        // only with chips in play (insurance is purely a money bet) that can
        // back the side bet.
        if betting, dealer.first?.isAce == true, canAffordInsurance {
            phase = .insurance
            return
        }
        peekAndContinue()
    }

    /// Check the hole card and continue the round: a dealer natural ends every
    /// box at once, an opening player natural is paid 3:2 and sits out, and the
    /// remaining boxes are played in order.
    private func peekAndContinue() {
        let dealer = dealerCards
        if Hand.isBlackjack(dealer) {
            // A dealer natural ends every box at once — no player action, no draw.
            for index in hands.indices {
                settleHand(at: index, dealer: dealer)
            }
            phase = .resolved
            return
        }
        for index in hands.indices where Hand.isBlackjack(hands[index].cards) {
            settleHand(at: index, dealer: dealer)
        }
        activateNextOrResolve()
    }

    /// Settle one hand against the dealer's final cards and record it. Idempotent:
    /// a hand that already carries a settlement (an opening natural) is left alone
    /// so its tally is never double-counted.
    private func settleHand(at index: Int, dealer: [Card]) {
        guard hands[index].settlement == nil else { return }
        let natural = hands[index].fromSplit ? false : Hand.isBlackjack(hands[index].cards)
        let result = Showdown.settle(
            player: hands[index].cards, dealer: dealer, playerNatural: natural
        )
        hands[index].settlement = result
        hands[index].done = true
        stats.record(outcome: result.outcome, playerBlackjack: result.playerBlackjack)
        if betting {
            let hand = hands[index]
            let chips = Bankroll.payout(settlement: result, bet: hand.bet, doubled: hand.doubled)
            bankrollStore.record(stake: stake(hand), payout: chips)
            roundNet += chips
        }
    }

    private func hit() {
        guard phase == .playerTurn, hands.indices.contains(activeIndex) else { return }
        guard let card = draw() else {
            finishActive()
            return
        }
        hands[activeIndex].cards.append(card)
        if Hand.isBust(hands[activeIndex].cards) { finishActive() }
    }

    private func stand() {
        guard phase == .playerTurn else { return }
        finishActive()
    }

    /// Double down: take exactly one card at a doubled stake, then the hand ends.
    private func double() {
        guard canDouble else { return }
        hands[activeIndex].doubled = true
        if let card = draw() { hands[activeIndex].cards.append(card) }
        finishActive()
    }

    /// Split a pair: the two cards seed two hands. The active hand keeps the first
    /// card and is dealt a new second card; the second card starts a new hand
    /// inserted right after, played once the active one finishes. Split aces take
    /// a single card each and stand.
    private func split() {
        guard canSplit else { return }
        let index = activeIndex
        let pair = hands[index].cards
        let box = hands[index].box
        let posted = hands[index].bet
        let splitAce = pair[0].isAce
        // Both halves stay in the box that split, so the box keeps its own cap, and
        // each carries the box's bet — a split posts a second one.
        hands.replaceSubrange(index ... index, with: [
            PlayerHand(
                cards: [pair[0]], box: box, isSplitAce: splitAce, fromSplit: true, bet: posted
            ),
            PlayerHand(
                cards: [pair[1]], box: box, isSplitAce: splitAce, fromSplit: true, bet: posted
            )
        ])
        dealToFreshHand(index)
    }

    /// Deal the second card to a one-card (freshly split) hand, then either finish
    /// it (a split ace stands after one card) or leave it for the player.
    private func dealToFreshHand(_ index: Int) {
        if let card = draw() { hands[index].cards.append(card) }
        let hand = hands[index]
        if hand.isSplitAce || hand.cards.count < 2 || Hand.isBust(hand.cards) {
            finishHand(index)
        }
    }

    private func finishActive() {
        finishHand(activeIndex)
    }

    private func finishHand(_ index: Int) {
        hands[index].done = true
        activateNextOrResolve()
    }

    /// Hand play to the earliest hand still owed a decision, or resolve the round
    /// when every hand is finished. Hands are always completed front to back, so
    /// the first not-done hand is the next one to act.
    private func activateNextOrResolve() {
        guard let next = hands.indices.first(where: { !hands[$0].done }) else {
            resolveAll()
            return
        }
        activeIndex = next
        // A freshly split hand arrives with one card; deal its second first.
        if hands[next].cards.count == 1 { dealToFreshHand(next) }
    }

    /// Reveal the dealer's hole card, play it out once (only if a hand can still
    /// win — every hand busted or already settled means no draw), then settle the
    /// hands still open. A split hand never counts as a natural, so its two-card
    /// 21 pays even money.
    private func resolveAll() {
        let anyLive = hands.contains { $0.settlement == nil && !Hand.isBust($0.cards) }
        if anyLive {
            dealerCards = Showdown.playDealerHand(dealerCards, ruleSet: ruleSet) { [weak self] in
                self?.draw()
            }
        }
        for index in hands.indices {
            settleHand(at: index, dealer: dealerCards)
        }
        phase = .resolved
    }

    private func draw() -> Card? {
        let dealt = shoe.deal(1)
        remaining = shoe.cardsRemaining
        if let card = dealt.first { dealtCards.append(card) }
        return dealt.first
    }

    func resetStats() {
        stats.reset()
    }
}

/// The insurance decision. Same file as the model — the mutators drive the
/// file-scoped `private(set)` state — but outside the class body to respect the
/// repo's type-length limit.
@MainActor
extension ShowdownModel {
    /// Insure every box for half its bet: the side bets settle against the hole
    /// card immediately — paid 2:1 on a dealer natural, forfeited otherwise —
    /// and the round then continues exactly as an uninsured one.
    func takeInsurance() {
        guard phase == .insurance else { return }
        let dealerBlackjack = Hand.isBlackjack(dealerCards)
        var net = 0.0
        for hand in hands {
            let payout = Bankroll.insurancePayout(bet: hand.bet, dealerBlackjack: dealerBlackjack)
            bankrollStore.record(stake: Bankroll.insuranceCost(bet: hand.bet), payout: payout)
            net += payout
        }
        insuranceNet = net
        roundNet += net
        phase = .playerTurn
        peekAndContinue()
    }

    func declineInsurance() {
        guard phase == .insurance else { return }
        phase = .playerTurn
        peekAndContinue()
    }
}
